ddl/db_integration_test.go

```diff
@@ -2927,3 +2927,116 @@ func (s *testIntegrationSuite3) TestAvoidCreateViewOnLocalTemporaryTable(c *C) {
+	defer tk.MustExec("drop table if exists a_normal_table")
+	tk.MustExec("show create table a_local_temp_table")
+	tk.MustExec("show create table a_local_temp_table_1")
+	tk.MustExec("show create table a_local_temp_table_2")
+	tk.MustExec("show create table a_normal_table")
+	tk.MustExec("drop table a_local_temp_table, a_local_temp_table_1, a_local_temp_table_2, a_normal_table")
+	tk.MustExec("rollback")
+	tk.MustQuery("select * from check_data_normal_table_1").Check(testkit.Rows("1"))
+
+	// Check drop not exists table.
+	tk.MustExec("create temporary table a_local_temp_table_3 (id int)")
+	tk.MustExec("create temporary table a_local_temp_table_4 (id int)")
+	tk.MustExec("create temporary table a_local_temp_table_5 (id int)")
+	tk.MustExec("drop table if exists a_normal_table_2")
+	tk.MustExec("create table a_normal_table_2 (id int)")
+	defer tk.MustExec("drop table if exists a_normal_table_2")
+	_, err = tk.Exec("drop table a_local_temp_table_3, a_local_temp_table_4, a_local_temp_table_5, a_normal_table_2, a_local_temp_table_6")
+	c.Assert(err.Error(), Equals, "[schema:1051]Unknown table 'test.a_local_temp_table_6'")
+	tk.MustQuery("select * from a_local_temp_table_3").Check(testkit.Rows())
+	tk.MustQuery("select * from a_local_temp_table_4").Check(testkit.Rows())
+	tk.MustQuery("select * from a_local_temp_table_5").Check(testkit.Rows())
+
+	tk.MustExec("drop table if exists check_data_normal_table_3")
+	tk.MustExec("create table check_data_normal_table_3 (id int)")
+	defer tk.MustExec("drop table if exists check_data_normal_table_3")
+	tk.MustExec("create temporary table a_local_temp_table_6 (id int)")
+	_, err = tk.Exec("drop table check_data_normal_table_3, check_data_normal_table_7, a_local_temp_table_6")
+	c.Assert(err.Error(), Equals, "[schema:1051]Unknown table 'test.check_data_normal_table_7'")
+	tk.MustQuery("select * from a_local_temp_table_6").Check(testkit.Rows())
+
+	// Check filter out data from removed local temp tables
+	tk.MustExec("create temporary table a_local_temp_table_7 (id int)")
+	ctx := s.ctx
+	c.Assert(ctx.NewTxn(context.Background()), IsNil)
+	txn, err := ctx.Txn(true)
+	c.Assert(err, IsNil)
+	defer func() {
+		err := txn.Rollback()
+		c.Assert(err, IsNil)
+	}()
+	sessionVars := tk.Se.GetSessionVars()
+	sessVarsTempTable := sessionVars.LocalTemporaryTables
+	localTemporaryTable := sessVarsTempTable.(*infoschema.LocalTemporaryTables)
+	tbl, exist := localTemporaryTable.TableByName(model.NewCIStr("test"), model.NewCIStr("a_local_temp_table_7"))
+	c.Assert(exist, IsTrue)
+	tblInfo := tbl.Meta()
+	tablePrefix := tablecodec.EncodeTablePrefix(tblInfo.ID)
+
+	tk.MustExec("begin")
+	tk.MustExec("insert into a_local_temp_table_7 values (1)")
+	tk.MustExec("drop table if exists a_local_temp_table_7")
+	tk.MustExec("commit")
+
+	_, err = tk.Exec("select * from a_local_temp_table_7")
+	c.Assert(err.Error(), Equals, "[schema:1146]Table 'test.a_local_temp_table_7' doesn't exist")
+	iter, _ := txn.Iter(tablePrefix, nil)
```

Comment:
I'm afraid you cannot get temporary session data from txn, it does not include local temporary data. You should iter `sessionvars.TemporaryTableData`